### PR TITLE
#67 Functions for unpacking based on byte arrays

### LINKED ISSUE #67: Unpack signal for CAN FD messages
If the dbc file contains CAN-FD messages that can be longer than 64 bit the unpack functions are not usable. 
Also i think most can implementations work by giving you a byte array for the received data anyway (peak, innodisk, esd to name a few).

Therefor i adjusted the unpack signals method like this:

```c#
    /// <summary>
    /// Get start bit Little Endian
    /// </summary>
    private static byte GetStartBitLE(Signal signal, int messageByteCount = 8)
    {
        var startByte = (byte)(signal.StartBit / 8);
        return (byte)(8 * messageByteCount - (signal.Length + 8 * startByte + (8 * (startByte + 1) - (signal.StartBit + 1)) % 8));
    }
```

And then adjust the code from before to 
```c#
    /// <summary>
    /// Function to unpack a signal from a CAN data message
    /// </summary>
    /// <param name="receiveMessage">The message data</param>
    /// <param name="signal">Signal containing dbc information</param>
    /// <returns>Returns a double value representing the unpacked signal</returns>
    private static double RxSignalUnpack(byte[] receiveMessage, Signal signal)
    {
        var bitMask = signal.BitMask();
        var startBit = signal.StartBit;
        
        if (!signal.Intel())
        {
            receiveMessage = receiveMessage.Reverse().ToArray();
            startBit = GetStartBitLE(signal, receiveMessage.Length);
        }
        
        // Unpack signal
        var receiveMessageBitArray = new BitArray(receiveMessage);
        receiveMessageBitArray.RightShift(startBit);
        receiveMessageBitArray.Length = 64; //Fill up to 64 if shorter or cut to length 64
        receiveMessageBitArray.And(new BitArray(BitConverter.GetBytes(bitMask)));
        
        var tempArray = new byte[8];
        receiveMessageBitArray.CopyTo(tempArray, 0);
        var iVal = BitConverter.ToInt64(tempArray, 0);

        // Manage sign bit (if signed)
        if (signal.ValueType == DbcValueType.Signed)
        {
            iVal -= ((iVal >> (signal.Length - 1)) != 0) ? (1L << signal.Length) : 0L;
        }
        else if (signal.ValueType == DbcValueType.IEEEFloat)
        {
            return (FloatConverter.AsFloatingPoint((int)iVal) * signal.Factor + signal.Offset);
        }
        else if (signal.ValueType == DbcValueType.IEEEDouble)
        {
            return (DoubleConverter.AsFloatingPoint(iVal) * signal.Factor + signal.Offset);
        }

        // Apply scaling
        return ((double)(iVal * (decimal)signal.Factor + (decimal)signal.Offset));
    }
```

this seems to work in a manual test. 
i tried to reuse most of the code. Probably is also not the fasted solution but should work.

~Problem with this is that the BitArray.ShiftRight() is not available in all targeted frameworks. For .net461 there is no solution for .netstandard2.0 you could upgrade the solution to .netstandard2.1.~

### PR DESCRIPTION
**Pull request by Uight**
This adds some new functions for packing/unpacking can messages based on byte arrays and does also support message lengths != 64 bit. 

For me this is considered a base implementation that could be used to build on. (E.g. Message encode/decode function)

I also added a new benchmark project to the solution using dotnetbenchmark. This is optional but could help to tune performance in the future.

resolves #67 (but could be extended in the future to implement message packing aso.)

### CODE REVIEW

**Comment by EFeru** [comment]
Hi @Uight ,
Thanks for the PR. You said there was something strange with the GetStartBitLE function. Did you manage to figure it out?
I am in parallel checking another implementation in js:

https://github.com/bit-dream/candied/blob/4e4077131d4104f417e750f916edae5cd9e76eb1/src/can/Can.ts#L167-L183

https://github.com/bit-dream/candied/blob/4e4077131d4104f417e750f916edae5cd9e76eb1/src/can/BitUtils.ts#L90-L102

**Comment by Uight** [comment]
@EFeru i figured that out i was just being stupid ;) 
i now did check everything against the python cantools as described in the issue and everything seems to be the same. Assuming its done right in the can tools its also done correctly here. 
If this PR goes through it would probably still be usefull to discuss how the final packer interface should look like in the end and do some more stuff there

**Comment by EFeru** [comment]
@Uight still need to look into this. I did not forget :)

File: DbcParserLib/Packer.cs [hunk 1]
```diff
@@ -99,14 +103,14 @@ public static double RxSignalUnpack(ulong RxMsg64, Signal signal)
...
+        public static double RxSignalUnpack(uint8_T[] receiveMessage, Signal signal)
         {
             var startBit = signal.StartBit;
             var message = receiveMessage;
 
             if (!signal.Intel())
             {
-                var tempArray = new byte[message.Length];
+                var tempArray = new uint8_T[message.Length];
                 Array.Copy(message, tempArray, message.Length);
```
Comment (Adhara3) [inline DbcParserLib/Packer.cs:106]: This part is bittersweet. It's easy and readable but allocating to just read backward it's a pity. 
Maybe if we change the `ExtractBits` signature into something like
```
private static uint64_T ExtractBits(uint8_T[] data, int startBit, int endBit, int step = 1) // Or -1 if backward
```
we could keep the same iteration without assuming that length is the limit. Just food for thoughts.

...or... (this is a bit outside the scope of this PR, but worth mentioning)
Today the `Packer` is a static class so it needs Signal and message info. If we optimize the `Signal` class (or if we adda a sort of `.GetReader()` on it) then we could optimize some stuff, removing ifs, and give a forward/backward reader depending on endianness

**Review by Adhara3 — APPROVED** [review]
Sounds good, just a minor comment.
Moreover some review has already been erformed with @EFeru and I see many UT, so we are ok.

Thanks

**Pull request merged**